### PR TITLE
Add test to train sharded resnet50 model

### PR DESCRIPTION
**Pull request by jonb377**
This is a draft PR and should not be merged

### CODE REVIEW

**Comment by jonb377** [comment]
Use the `--sharded` flag to apply SPMD to the model weights, e.g. `python test/test_train_spmd_imagenet.py --fake_data --sharded`

File: torch_xla/csrc/tensor.cpp [hunk 1]
```diff
@@ -705,7 +705,7 @@ void XLATensor::AssignIrValue(torch::lazy::Value ir_value) const {
   // TODO(yeounoh): Sharding annotation must be removed by explicit call to
   // ClearSharding.
   ShardingSpecPtr sharding = sharding_spec();
-  if (sharding != nullptr) {
+  if (ir_value && sharding != nullptr) {
```
Comment (yeounoh) [inline torch_xla/csrc/tensor.cpp:708]: Let's rebase after merging #4233

File: test/test_train_spmd_imagenet.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+  torch.manual_seed(42)
+
+  device = xm.xla_device()
+  model = torchvision.models.resnet50().to(device)
+  if FLAGS.sharded:
+    for name, layer in model.named_modules():
+      if 'conv' in name:
+        print('Attempting to apply sharding to', name)
+        partition = int(xm.xrt_world_size() / 2)
+        xs.mark_sharding(layer.weight, (1, 1, partition, partition), (0, 1, 2, 3))
```
Comment (yeounoh) [inline test/test_train_spmd_imagenet.py:155]: Rebase and use `Mesh`

File: third_party/xla_client/BUILD [hunk 3]
```diff
@@ -47,8 +47,10 @@ tf_cc_shared_object(
             "-z defs",
             "-Wl,--version-script",  #  This line must be directly followed by the version_script.lds file
             "$(location //tensorflow/compiler/xla/xla_client:tf_version_script.lds)",
+            "-g",
```
Comment (yeounoh) [inline third_party/xla_client/BUILD:50]: Are they still relevant in this PR? This was for debug builds?

**Review by yeounoh — CHANGES_REQUESTED** [review]
Thanks, left a few comments -- otherwise mostly looks good to me.

File: test/spmd/test_train_spmd_imagenet.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+import torch_xla
+import torch_xla.debug.metrics as met
+import torch_xla.distributed.parallel_loader as pl
+import torch_xla.utils.utils as xu
+import torch_xla.core.xla_model as xm
+import torch_xla.debug.profiler as xp
+import torch_xla.test.test_utils as test_utils
+import torch_xla.experimental.xla_sharding as xs
+from contextlib import nullcontext
+import torch_xla.debug.metrics as met
```
Comment (yeounoh) [inline test/spmd/test_train_spmd_imagenet.py:72]: Nit. Duplicate.

File: torch_patches/.torch_pin [hunk 5]
```diff
@@ -0,0 +1 @@
+31708a731076b7feed3051b81d309a9babb4efc0
```
Comment (yeounoh) [inline torch_patches/.torch_pin:1]: Rebase and remove this before merging.

File: third_party/xla_client/BUILD (hunk 3 above)
Comment (jonb377) [reply third_party/xla_client/BUILD]: Right, this was for debugging. I've removed it from the later revisions.

File: test/spmd/test_train_spmd_imagenet.py (hunk 4 above)
Comment (jonb377) [reply test/spmd/test_train_spmd_imagenet.py]: Good catch, thanks!

File: test/spmd/test_train_spmd_imagenet.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+def train_imagenet():
+  print('==> Preparing data..')
+  img_dim = get_model_property('img_dim')
+  if FLAGS.fake_data:
+    train_dataset_len = 1200000  # Roughly the size of Imagenet dataset.
+    train_loader = xu.SampleGenerator(
+        data=(torch.zeros(FLAGS.batch_size, 3, img_dim, img_dim),
+              torch.zeros(FLAGS.batch_size, dtype=torch.int64)),
+        sample_count=train_dataset_len // FLAGS.batch_size //
+        xm.xrt_world_size())
```
Comment (yeounoh) [inline test/spmd/test_train_spmd_imagenet.py:137]: You don't need `// xm.xrt_world_size()`

**Review by yeounoh — APPROVED** [review]
LGTM

**Pull request merged**